src/Arcus.WebApi.Security.Authentication/SharedAccessKey/SharedAccessKeyAuthenticationFilter.cs

```diff
@@ -53,35 +55,55 @@ public async Task OnAuthorizationAsync(AuthorizationFilterContext context)
             Guard.For<ArgumentException>(() => context.HttpContext.Request.Headers == null, "Invalid action context given without any HTTP request headers");
             Guard.For<ArgumentException>(() => context.HttpContext.RequestServices == null, "Invalid action context given without any HTTP request services");
 
-            if (context.HttpContext.Request.Headers
-                       .TryGetValue(_headerName, out StringValues requestSecretHeaders))
+            string foundSecret = await GetAuthorizationSecret(context);
+
+            if (!context.HttpContext.Request.Headers.ContainsKey(_headerName) && !context.HttpContext.Request.Query.ContainsKey(_queryParameterName))
+            {
+                context.Result = new UnauthorizedResult();
+            }
+            else
             {
-                ISecretProvider userDefinedSecretProvider = 
-                    context.HttpContext.RequestServices.GetService<ICachedSecretProvider>()
-                    ?? context.HttpContext.RequestServices.GetService<ISecretProvider>();
-                
-                if (userDefinedSecretProvider == null)
+                if (!String.IsNullOrWhiteSpace(_headerName) && context.HttpContext.Request.Headers
+                       .TryGetValue(_headerName, out StringValues requestSecretHeaders))
                 {
-                    throw new KeyNotFoundException(
-                        $"No configured {nameof(ICachedSecretProvider)} or {nameof(ISecretProvider)} implementation found in the request service container. "
-                        + "Please configure such an implementation (ex. in the Startup) of your application");
-                }
 
-                string foundSecret = await userDefinedSecretProvider.Get(_secretName);
-                if (foundSecret == null)
-                {
-                    throw new SecretNotFoundException(_secretName);
+                    if (requestSecretHeaders.Any(headerValue => !String.Equals(headerValue, foundSecret)))
```

Comment:
Feels like this is easier to read, no? 
```suggestion
                    if (requestSecretHeaders.Any(headerValue => string.Equals(headerValue, foundSecret, StringComparison.InvariantCultureIgnoreCase)) == false)
```

We should also be case insensitive, etc. - Maybe another good candidate for a test?

Reply:
Aren't SharedAccessKey ment to be case-sensitive?

Reply:
You are correct, sorry about that!